### PR TITLE
Pipeline Concurrency

### PR DESCRIPTION
**Pull request by hutuxian**
Add the pipeline concurrency function

### CODE REVIEW

File: paddle/fluid/framework/data_feed.cc [hunk 1]
```diff
@@ -980,5 +970,166 @@ void MultiSlotInMemoryDataFeed::DeserializeIns(
...
+      int64_t* tensor_ptr = feed_vec_[i]->mutable_data<int64_t>(
+          {total_instance, 1}, platform::CPUPlace());
+      memcpy(tensor_ptr, &feasign[0], total_instance * sizeof(int64_t));
+    }
+
+    LoD data_lod{offset};
+    feed_vec_[i]->set_lod(data_lod);
+    if (use_slots_is_dense_[i]) {
+      int dim = total_instance / batch_size_;
+      feed_vec_[i]->Resize({batch_size_, dim});
```
Comment (guru4elephant) [inline paddle/fluid/framework/data_feed.cc]: I've fixed some bugs of dense tensor Resize, maybe you need to adapt the changes later.

File: paddle/fluid/framework/data_feed.cc [hunk 2]
```diff
@@ -762,8 +762,8 @@ void MultiSlotDataFeed::PutToFeedVec(
     LoD data_lod{offset};
     feed_vec_[i]->set_lod(data_lod);
     if (use_slots_is_dense_[i]) {
-      use_slots_shape_[i][0] = batch_size_;
-      feed_vec_[i]->Resize(framework::make_ddim(use_slots_shape_[i]));
+      int dim = total_instance / batch_size_;
```
Comment (guru4elephant) [inline paddle/fluid/framework/data_feed.cc]: I fixed some bugs here, maybe you can merge from develop.

File: paddle/fluid/framework/data_feed.h [hunk 3]
```diff
@@ -253,16 +254,25 @@ class MultiSlotType {
  public:
   MultiSlotType() {}
   ~MultiSlotType() {}
-  void Init(const std::string& type) {
+  void Init(const std::string& type, size_t reserved_size=0) {
```
Comment (guru4elephant) [inline paddle/fluid/framework/data_feed.h]: why do you have to add reserved_size?

File: paddle/fluid/framework/device_worker.h [hunk 4]
```diff
@@ -34,10 +35,25 @@ limitations under the License. */
...
 #include "paddle/fluid/platform/timer.h"
+#include "paddle/fluid/platform/nccl_helper.h"
 
 namespace paddle {
 namespace framework {
 
+//#define DEBUG
+
+#ifdef DEBUG
+#define HTJ LOG(ERROR) << __FUNCTION__ << ": "
```
Comment (guru4elephant) [inline paddle/fluid/framework/device_worker.h]: what is HTJ?

File: paddle/fluid/framework/executor.cc [hunk 5]
```diff
@@ -122,8 +122,8 @@ void Executor::RunFromDataset(const ProgramDesc& main_program, Scope* scope,
                               const std::string& trainer_desc_str) {
   VLOG(3) << "Start to RunFromDataset in executor";
   TrainerDesc trainer_desc;
-  google::protobuf::TextFormat::ParseFromString(trainer_desc_str,
-                                                &trainer_desc);
+  bool success = trainer_desc.ParseFromString(trainer_desc_str);
```
Comment (guru4elephant) [inline paddle/fluid/framework/executor.cc:125]: great, thanks for your fix.

File: paddle/fluid/framework/trainer.h [hunk 6]
```diff
@@ -91,5 +91,47 @@ class DistMultiTrainer : public MultiTrainer {
...
+  PipelineWorkerParameter pipeline_config_;
+  
+  std::vector<std::unique_ptr<std::vector<std::string>>> in_var_names_;
+  std::vector<std::unique_ptr<std::vector<std::string>>> out_var_names_;
+
+  std::vector<std::vector<std::unique_ptr<std::mutex>>> worker_count_mutex_;
+  std::vector<std::vector<int*>> worker_count_;
+
+  // worker: [section_id][line_id][id]
+  std::vector<std::vector<std::vector<std::shared_ptr<paddle::framework::DeviceWorker>>>> workers_;
```
Comment (guru4elephant) [inline paddle/fluid/framework/trainer.h]: workers_ here should be documented more clearly.

File: paddle/fluid/framework/trainer.h [hunk 7]
```diff
@@ -91,5 +91,47 @@ class DistMultiTrainer : public MultiTrainer {
...
+  std::vector<std::thread> section_threads_;
+  
+  std::vector<std::vector<std::unique_ptr<ScopeQueue>>> scope_queues_;
```
Comment (guru4elephant) [inline paddle/fluid/framework/trainer.h:130]: scope_queues_ should be documented clearly, this is a very important data structure in your design.

File: python/paddle/fluid/executor.py [hunk 8]
```diff
@@ -637,12 +637,18 @@ def _prepare_trainer(self,
         assert len(fetch_list) == len(fetch_info)
         compiled = isinstance(program, compiler.CompiledProgram)
         if not compiled:
-            trainer = TrainerFactory()._create_trainer(program._fleet_opt)
+            if program._pipeline_opt:
```
Comment (guru4elephant) [inline python/paddle/fluid/executor.py:785]: not sure whether this matches the newest API, please check and merge diff.

**Review by gavin1332 — APPROVED** [review]
code review have be done in the offline

**Review by xsrobin — APPROVED** [review]
LGTM

**Review by luotao1 — APPROVED** [review]
LG for api.spec

**Pull request merged**